File: circuitUI/1/Circuit.cs
using System;
using System.Collections.Generic;
using System.Numerics;

namespace CircuitSimulator
{
    public enum AnalysisType
    {
        DC,
        TRANSIENT,
        AC_SWEEP
    }

    public class Circuit
    {
        public List<Node> Nodes { get; private set; }
        public List<Resistor> Resistors { get; private set; }
        public List<Capacitor> Capacitors { get; private set; }
        public List<Inductor> Inductors { get; private set; }
        public List<Diode> Diodes { get; private set; }
        public List<VoltageSource> VoltageSources { get; private set; }
        public List<ACVoltageSource> AcVoltageSources { get; private set; }
        public List<CurrentSource> CurrentSources { get; private set; }
        public List<string> GroundNodeNames { get; private set; }

        public double DeltaT { get; set; }

        public List<List<double>> MNA_A { get; private set; }
        public List<double> MNA_RHS { get; private set; }

        public List<List<Complex>> MNA_A_Complex { get; private set; }
        public List<Complex> MNA_RHS_Complex { get; private set; }

        public Circuit()
        {
            Nodes = new List<Node>();
            Resistors = new List<Resistor>();
            Capacitors = new List<Capacitor>();
            Inductors = new List<Inductor>();
            Diodes = new List<Diode>();
            VoltageSources = new List<VoltageSource>();
            AcVoltageSources = new List<ACVoltageSource>();
            CurrentSources = new List<CurrentSource>();
            GroundNodeNames = new List<string>();

            DeltaT = 0;

            MNA_A = new List<List<double>>();
            MNA_RHS = new List<double>();

            MNA_A_Complex = new List<List<Complex>>();
            MNA_RHS_Complex = new List<Complex>();
        }

        public void AddNode(string name)
        {
            if (FindNode(name) == null)
            {
                Node newNode = new Node();
                newNode.Name = name;
                Nodes.Add(newNode);
            }
        }

        public Node FindNode(string name)
        {
            foreach (Node node in Nodes)
            {
                if (node.Name == name)
                {
                    return node;
                }
            }
            return null;
        }

        public Node FindOrCreateNode(string name)
        {
            Node node = FindNode(name);
            if (node != null)
            {
                return node;
            }
            AddNode(name);
            return Nodes[Nodes.Count - 1];
        }

        public Resistor FindResistor(string name)
        {
            foreach (Resistor res in Resistors)
            {
                if (res.Name == name)
                    return res;
            }
            return null;
        }

        public Capacitor FindCapacitor(string name)
        {
            foreach (Capacitor cap in Capacitors)
            {
                if (cap.Name == name)
                    return cap;
            }
            return null;
        }

        public Inductor FindInductor(string name)
        {
            foreach (Inductor ind in Inductors)
            {
                if (ind.Name == name)
                    return ind;
            }
            return null;
        }

        public Diode FindDiode(string name)
        {
            foreach (Diode d in Diodes)
            {
                if (d.Name == name)
                    return d;
            }
            return null;
        }

        public CurrentSource FindCurrentSource(string name)
        {
            foreach (CurrentSource cs in CurrentSources)
            {
                if (cs.Name == name)
                    return cs;
            }
            return null;
        }

        public VoltageSource FindVoltageSource(string name)
        {
            foreach (VoltageSource vs in VoltageSources)
            {
                if (vs.Name == name)
                    return vs;
            }
            return null;
        }

        public ACVoltageSource FindACVoltageSource(string name)
        {
            foreach (ACVoltageSource acVs in AcVoltageSources)
            {
                if (acVs.Name == name)
                    return acVs;
            }
            return null;
        }

        public bool DeleteResistor(string name)
        {
            Resistor res = FindResistor(name);
            if (res != null)
            {
                Resistors.Remove(res);
                return true;
            }
            return false;
        }

        public bool DeleteCapacitor(string name)
        {
            Capacitor cap = FindCapacitor(name);
            if (cap != null)
            {
                Capacitors.Remove(cap);
                return true;
            }
            return false;
        }

        public bool DeleteInductor(string name)
        {
            Inductor ind = FindInductor(name);
            if (ind != null)
            {
                Inductors.Remove(ind);
                return true;
            }
            return false;
        }

        public bool DeleteDiode(string name)
        {
            Diode diode = FindDiode(name);
            if (diode != null)
            {
                Diodes.Remove(diode);
                return true;
            }
            return false;
        }

        public bool DeleteVoltageSource(string name)
        {
            VoltageSource vs = FindVoltageSource(name);
            if (vs != null)
            {
                VoltageSources.Remove(vs);
                return true;
            }
            return false;
        }

        public bool DeleteCurrentSource(string name)
        {
            CurrentSource cs = FindCurrentSource(name);
            if (cs != null)
            {
                CurrentSources.Remove(cs);
                return true;
            }
            return false;
        }

        public int CountTotalExtraVariables()
        {
            int mVars = VoltageSources.Count + Inductors.Count;
            foreach (Diode diode in Diodes)
            {
                if (diode.GetState() == DiodeState.STATE_FORWARD_ON || diode.GetState() == DiodeState.STATE_REVERSE_ON)
                {
                    mVars++;
                }
            }
            return mVars;
        }

        public void AssignDiodeBranchIndices()
        {
            int currentBranchIdx = VoltageSources.Count + Inductors.Count;
            foreach (Diode diode in Diodes)
            {
                if (diode.GetState() == DiodeState.STATE_FORWARD_ON || diode.GetState() == DiodeState.STATE_REVERSE_ON)
                {
                    diode.SetBranchIndex(currentBranchIdx++);
                }
                else
                {
                    diode.SetBranchIndex(-1);
                }
            }
        }

        public void SetDeltaT(double dt)
        {
            this.DeltaT = dt;
        }

        public void UpdateComponentStates()
        {
            foreach (Capacitor cap in Capacitors)
            {
                cap.Update(DeltaT);
            }
            foreach (Inductor ind in Inductors)
            {
                ind.Update(DeltaT);
            }
        }

        public void ClearComponentHistory()
        {
            foreach (Node node in Nodes)
            {
                node.ClearHistory();
            }
            foreach (VoltageSource vs in VoltageSources)
            {
                vs.ClearHistory();
            }
        }

        public int GetNodeMatrixIndex(Node targetNodePtr)
        {
            if (targetNodePtr == null || targetNodePtr.IsGround)
            {
                return -1;
            }
            int matrixIdx = 0;
            foreach (Node nInList in Nodes)
            {
                if (!nInList.IsGround)
                {
                    if (nInList.Num == targetNodePtr.Num)
                    {
                        return matrixIdx;
                    }
                    matrixIdx++;
                }
            }
            return -1;
        }

        public int CountNonGroundNodes()
        {
            int count = 0;
            foreach (Node node in Nodes)
            {
                if (!node.IsGround)
                {
                    count++;
                }
            }
            return count;
        }
    }
}

File: src/Circuit.cpp
#include "Circuit.h"
#include <algorithm>
#include <vector>
#include <string>
#include <complex>

Circuit::Circuit() : delta_t(0), currentTime(0) {}

Circuit::~Circuit() {
    for (Node* node : nodes) {
        delete node;
    }
    nodes.clear();
}

void Circuit::addNode(const std::string& name) {
    if (!findNode(name)) {
        Node* newNode = new Node();
        newNode->name = name;
        nodes.push_back(newNode);
    }
}

Node* Circuit::findNode(const std::string& find_from_name) {
    for (Node* node : nodes) {
        if (node->name == find_from_name) {
            return node;
        }
    }
    return nullptr;
}


Node* Circuit::findOrCreateNode(const std::string& name) {
    Node* node = findNode(name);
    if (node) {
        return node;
    }
    addNode(name);
    return nodes.back();
}

Resistor* Circuit::findResistor(const std::string& find_from_name) {
    for (auto& res : resistors) { if (res.name == find_from_name) return &res; }
    return nullptr;
}

Capacitor* Circuit::findCapacitor(const std::string& find_from_name) {
    for (auto& cap : capacitors) { if (cap.name == find_from_name) return &cap; }
    return nullptr;
}

Inductor* Circuit::findInductor(const std::string& find_from_name) {
    for (auto& ind : inductors) { if (ind.name == find_from_name) return &ind; }
    return nullptr;
}

VoltageSource* Circuit::findVoltageSource(const std::string& find_from_name) {
    for (auto& vs : voltageSources) { if (vs.name == find_from_name) return &vs; }
    return nullptr;
}

ACVoltageSource* Circuit::findACVoltageSource(const std::string& find_from_name) {
    for (auto& ac_vs : acVoltageSources) { if (ac_vs.name == find_from_name) return &ac_vs; }
    return nullptr;
}

CurrentSource* Circuit::findCurrentSource(const std::string& find_from_name) {
    for (auto& cs : currentSources) { if (cs.name == find_from_name) return &cs; }
    return nullptr;
}

Diode* Circuit::findDiode(const std::string& find_from_name) {
    for (auto& d : diodes) { if (d.name == find_from_name) return &d; }
    return nullptr;
}


int Circuit::findNodeIndex(const std::string& name) {
    for (size_t i = 0; i < nodes.size(); ++i) {
        if (nodes[i]->name == name) {
            return static_cast<int>(i);
        }
    }
    return -1;
}

int Circuit::findVoltageSourceIndex(const std::string& name) {
    for (size_t i = 0; i < voltageSources.size(); ++i) {
        if (voltageSources[i].name == name) {
            return static_cast<int>(i);
        }
    }
    return -1;
}

// --- NEW HELPER FUNCTION ---
// Added a helper to find the index of an AC voltage source
int Circuit::findACVoltageSourceIndex(const std::string& name) {
    for (size_t i = 0; i < acVoltageSources.size(); ++i) {
        if (acVoltageSources[i].name == name) {
            return static_cast<int>(i);
        }
    }
    return -1;
}


void Circuit::setGroundNode(const std::string& node_name) {
    Node* node = findNode(node_name);
    if (node) {
        node->isGround = true;
    }
}

int Circuit::getNodeMatrixIndex(const Node* target_node_ptr) const {
    int non_gnd_idx_counter = 0;
    for (const auto& current_node_ptr : nodes) {
        if (current_node_ptr == target_node_ptr) {
            return non_gnd_idx_counter;
        }
        if (!current_node_ptr->isGround) {
            non_gnd_idx_counter++;
        }
    }
    return -1; // Should not happen if node exists
}

void Circuit::assignDiodeBranchIndices() {
    int n_nodes = 0;
    for (const auto& node : nodes) {
        if (!node->isGround) n_nodes++;
    }
    int branch_idx_offset = n_nodes + voltageSources.size() + inductors.size();
    for (auto& diode : diodes) {
        if (diode.getState() != STATE_OFF) {
            diode.setBranchIndex(branch_idx_offset++);
        }
        else {
            diode.setBranchIndex(-1);
        }
    }
}


void Circuit::set_MNA_A(AnalysisType type, double frequency) {
    int n = 0;
    for (const auto& node : nodes) {
        if (!node->isGround) n++;
    }

    // --- MODIFIED ---
    // AC Voltage Sources also need a branch current, so they are added to 'm'
    int m = voltageSources.size() + acVoltageSources.size() + inductors.size();
    int active_diodes = 0;
    if (type == AnalysisType::DC) {
        for (const auto& diode : diodes) {
            if (diode.getState() != STATE_OFF) {
                active_diodes++;
            }
        }
    }

    int matrix_size = n + m + active_diodes;
    MNA_A.assign(matrix_size, std::vector<double>(matrix_size, 0.0));
    MNA_A_Complex.assign(matrix_size, std::vector<std::complex<double>>(matrix_size, { 0.0, 0.0 }));

    if (type == AnalysisType::DC || type == AnalysisType::TRANSIENT) {
        // G Matrix (Resistors)
        for (const auto& res : resistors) {
            int n1_idx = getNodeMatrixIndex(res.node1);
            int n2_idx = getNodeMatrixIndex(res.node2);
            double g = 1.0 / res.resistance;
            if (n1_idx != -1) MNA_A[n1_idx][n1_idx] += g;
            if (n2_idx != -1) MNA_A[n2_idx][n2_idx] += g;
            if (n1_idx != -1 && n2_idx != -1) {
                MNA_A[n1_idx][n2_idx] -= g;
                MNA_A[n2_idx][n1_idx] -= g;
            }
        }
        if (type == AnalysisType::TRANSIENT) {
            // Add capacitor conductance for transient
            for (const auto& cap : capacitors) {
                int n1_idx = getNodeMatrixIndex(cap.node1);
                int n2_idx = getNodeMatrixIndex(cap.node2);
                double g_c = cap.capacitance / delta_t;
                if (n1_idx != -1) MNA_A[n1_idx][n1_idx] += g_c;
                if (n2_idx != -1) MNA_A[n2_idx][n2_idx] += g_c;
                if (n1_idx != -1 && n2_idx != -1) {
                    MNA_A[n1_idx][n2_idx] -= g_c;
                    MNA_A[n2_idx][n1_idx] -= g_c;
                }
            }
        }

        // B & C Matrix (Voltage Sources & Inductors)
        // DC Voltage Sources
        for (size_t i = 0; i < voltageSources.size(); ++i) {
            int n1_idx = getNodeMatrixIndex(voltageSources[i].node1);
            int n2_idx = getNodeMatrixIndex(voltageSources[i].node2);
            int branch_idx = n + i;
            if (n1_idx != -1) { MNA_A[n1_idx][branch_idx] += 1.0; MNA_A[branch_idx][n1_idx] += 1.0; }
            if (n2_idx != -1) { MNA_A[n2_idx][branch_idx] -= 1.0; MNA_A[branch_idx][n2_idx] -= 1.0; }
        }
        
        // --- NEW ---
        // AC Voltage Sources (for transient)
        int vs_offset = voltageSources.size();
        for (size_t i = 0; i < acVoltageSources.size(); ++i) {
            int n1_idx = getNodeMatrixIndex(acVoltageSources[i].node1);
            int n2_idx = getNodeMatrixIndex(acVoltageSources[i].node2);
            int branch_idx = n + vs_offset + i;
            if (n1_idx != -1) { MNA_A[n1_idx][branch_idx] += 1.0; MNA_A[branch_idx][n1_idx] += 1.0; }
            if (n2_idx != -1) { MNA_A[n2_idx][branch_idx] -= 1.0; MNA_A[branch_idx][n2_idx] -= 1.0; }
        }

        // Inductors
        int acvs_offset = vs_offset + acVoltageSources.size();
        for (size_t i = 0; i < inductors.size(); ++i) {
            int n1_idx = getNodeMatrixIndex(inductors[i].node1);
            int n2_idx = getNodeMatrixIndex(inductors[i].node2);
            int branch_idx = n + acvs_offset + i;
            if (n1_idx != -1) { MNA_A[n1_idx][branch_idx] += 1.0; MNA_A[branch_idx][n1_idx] += 1.0; }
            if (n2_idx != -1) { MNA_A[n2_idx][branch_idx] -= 1.0; MNA_A[branch_idx][n2_idx] -= 1.0; }
            if (type == AnalysisType::TRANSIENT) {
                 MNA_A[branch_idx][branch_idx] -= inductors[i].inductance / delta_t;
            }
        }
    }
    else if (type == AnalysisType::AC_SWEEP) {
        // Handle complex AC analysis...
    }
}

void Circuit::set_MNA_RHS(AnalysisType type, double frequency) {
    int n = 0;
    for (const auto& node : nodes) { if (!node->isGround) n++; }

    // --- MODIFIED ---
    // M must include ALL components with a branch current equation.
    int m = voltageSources.size() + acVoltageSources.size() + inductors.size();
    
    int active_diodes = 0;
    if (type == AnalysisType::DC) {
        for (const auto& diode : diodes) {
            if (diode.getState() != STATE_OFF) active_diodes++;
        }
    }

    int rhs_size = n + m + active_diodes;
    MNA_RHS.assign(rhs_size, 0.0);
    MNA_RHS_Complex.assign(rhs_size, { 0.0, 0.0 });
    
    // I vector (current sources)
    for (const auto& cs : currentSources) {
        int n1_idx = getNodeMatrixIndex(cs.node1);
        int n2_idx = getNodeMatrixIndex(cs.node2);
        if (n1_idx != -1) MNA_RHS[n1_idx] -= cs.value;
        if (n2_idx != -1) MNA_RHS[n2_idx] += cs.value;
    }
    if (type == AnalysisType::TRANSIENT) {
        // Capacitor current sources for transient
        for (const auto& cap : capacitors) {
            int n1_idx = getNodeMatrixIndex(cap.node1);
            int n2_idx = getNodeMatrixIndex(cap.node2);
            double i_c = (cap.capacitance / delta_t) * cap.prevVoltage;
            if (n1_idx != -1) MNA_RHS[n1_idx] += i_c;
            if (n2_idx != -1) MNA_RHS[n2_idx] -= i_c;
        }
        // Inductor voltage sources for transient
        int acvs_offset = voltageSources.size() + acVoltageSources.size();
        for (size_t i = 0; i < inductors.size(); ++i) {
            int branch_idx = n + acvs_offset + i;
            MNA_RHS[branch_idx] -= (inductors[i].inductance / delta_t) * inductors[i].prevCurrent;
        }
    }

    // E vector (voltage sources)
    // DC sources
    for (size_t i = 0; i < voltageSources.size(); ++i) {
        MNA_RHS[n + i] = voltageSources[i].value;
    }

    // --- NEW ---
    // AC sources in transient analysis
    if (type == AnalysisType::TRANSIENT) {
        int vs_offset = voltageSources.size();
        for (size_t i = 0; i < acVoltageSources.size(); ++i) {
            MNA_RHS[n + vs_offset + i] = acVoltageSources[i].getValue(currentTime);
        }
    }
}


void Circuit::MNA_sol_size() {
    MNA_solution.resize(MNA_A.size());
}

void Circuit::setDeltaT(double dt) {
    this->delta_t = dt;
}

void Circuit::updateComponentStates() {
    for (auto& cap : capacitors) {
        cap.update(delta_t);
    }
    for (auto& ind : inductors) {
        ind.update(delta_t);
    }
}

void Circuit::clearComponentHistory() {
    for (Node* node : nodes) {
        node->clearHistory();
    }
    // Also clear component histories
    for (auto& r : resistors) r.clearHistory();
    for (auto& c : capacitors) c.clearHistory();
    for (auto& l : inductors) l.clearHistory();
    for (auto& vs : voltageSources) vs.clearHistory();
    for (auto& ac_vs : acVoltageSources) ac_vs.clearHistory();
}

bool Circuit::isNodeNameGround(const std::string& node_name) const {
    for (const auto& gnd_name : groundNodeNames) {
        if (gnd_name == node_name) {
            return true;
        }
    }
    return false;
}

